[WPE][GTK] Speech synthesis using libspiel

philn opened the pull request:
#### 10381a8de938024f31935184af470ccbce0b22d9
<pre>
[WPE][GTK] Speech synthesis using libspiel
<a href="https://bugs.webkit.org/show_bug.cgi?id=270742">https://bugs.webkit.org/show_bug.cgi?id=270742</a>

Reviewed by Xabier Rodriguez-Calvar and Miguel Gomez.

Spiel&apos;s API is much nicer than Flite&apos;s, being directly inspired from WebSpeech. This new backend can
be enabled by setting USE_SPIEL=ON and USE_FLITE=OFF with cmake. The flite backend remains enabled
by default until Spiel becomes widly available in the various distributions.

* Source/WebCore/PlatformGTK.cmake:
* Source/WebCore/PlatformWPE.cmake:
* Source/WebCore/platform/PlatformSpeechSynthesizer.h:
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::registerWebKitGStreamerElements):
* Source/WebCore/platform/gstreamer/GUniquePtrFlite.h:
* Source/WebCore/platform/gstreamer/PlatformSpeechSynthesizerGStreamer.cpp:
* Source/WebCore/platform/gstreamer/WebKitFliteSourceGStreamer.cpp:
* Source/WebCore/platform/gstreamer/WebKitFliteSourceGStreamer.h:
* Source/WebCore/platform/spiel/PlatformSpeechSynthesizerSpiel.cpp: Added.
(WebCore::SpielSpeechWrapper::SpielSpeechWrapper):
(WebCore::SpielSpeechWrapper::finishSpeakerInitialization):
(WebCore::SpielSpeechWrapper::~SpielSpeechWrapper):
(WebCore::SpielSpeechWrapper::generateVoiceURI):
(WebCore::SpielSpeechWrapper::initializeVoiceList):
(WebCore::SpielSpeechWrapper::pause):
(WebCore::SpielSpeechWrapper::resume):
(WebCore::SpielSpeechWrapper::speakUtterance):
(WebCore::SpielSpeechWrapper::cancel):
(WebCore::SpielSpeechWrapper::clearUtterance):
(WebCore::PlatformSpeechSynthesizer::create):
(WebCore::PlatformSpeechSynthesizer::PlatformSpeechSynthesizer):
(WebCore::PlatformSpeechSynthesizer::initializeVoiceList):
(WebCore::PlatformSpeechSynthesizer::pause):
(WebCore::PlatformSpeechSynthesizer::resume):
(WebCore::PlatformSpeechSynthesizer::speak):
(WebCore::PlatformSpeechSynthesizer::cancel):
(WebCore::PlatformSpeechSynthesizer::resetState):
* Source/cmake/FindLibSpiel.cmake: Added.
* Source/cmake/GStreamerDependencies.cmake:
* Source/cmake/OptionsGTK.cmake:
* Source/cmake/OptionsWPE.cmake:

Canonical link: <a href="https://commits.webkit.org/285891@main">https://commits.webkit.org/285891@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e97b23a3b5dd15db71ab810c0662186d6c55f9f0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73943 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53372 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26754 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78291 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25181 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/76060 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62505 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1157 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58144 "Found 81 new test failures: accessibility/w3c-svg-content-language-attribute.html fast/css/focus-ring-exists-for-search-field.html fast/inline/list-marker-inside-container-with-margin.html fast/inline/min-content-width-with-hypens.html fast/inline/overflowing-content-with-hypens.html fast/mediastream/canvas-video-to-canvas.html fast/mediastream/captureStream/canvas3d.html fast/multicol/table-vertical-align.html fast/ruby/annotation-with-line-gap.html fast/ruby/ruby-with-continuation-crash.html ... (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16491 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77010 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48284 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63607 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38544 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/45111 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21097 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23514 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/67080 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66653 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21445 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79834 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/73201 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1260 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/659 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66460 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1403 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63621 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65741 "Found 1 new API test failure: /TestWebKit:WebKit.InjectedBundleInitializationUserDataCallbackWins (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16325 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9633 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7812 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/94982 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1224 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/4060 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20869 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1253 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/1240 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1259 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

magomez commented on Source/WebCore/PlatformGTK.cmake:
```diff
@@ -120,7 +120,17 @@ if (ENABLE_BUBBLEWRAP_SANDBOX)
     list(APPEND WebCore_LIBRARIES Libseccomp::Libseccomp)
 endif ()
 
-if (ENABLE_SPEECH_SYNTHESIS)
```
If SPEECH_SYNTHESIS is disabled but spiel is set to on, removing this check causes that the spiel bits get added to the build. Eihter a depend is added to spiel so it's disabled when SPEECH_SYNTHESIS gets disabled, or we keep this check here.

magomez commented on Source/WebCore/PlatformWPE.cmake:
```diff
@@ -141,7 +141,17 @@ if (ENABLE_GAMEPAD)
     )
 endif ()
 
-if (ENABLE_SPEECH_SYNTHESIS)
```
ditto

magomez commented on Source/WebCore/platform/spiel/PlatformSpeechSynthesizerSpiel.cpp:
```diff
@@ -0,0 +1,308 @@
+#include "config.h"
+#include "PlatformSpeechSynthesizer.h"
+
+#if ENABLE(SPEECH_SYNTHESIS) && USE(SPIEL)
```
In other places you seem to be moving to check USE(SPIEL) or USE(FLITE) and not check whether speech synthesis is enabled. But here you're checking ENABLE(SPEECH_SYNTHESIS) as well. Why? You should fix the depend problem of spiel and then you can stop checking for speech synthesis completely.

magomez commented on Source/cmake/OptionsGTK.cmake:
```diff
@@ -357,9 +359,20 @@ if (USE_GBM)
+        if (NOT LibSpiel_FOUND)
+            message(FATAL_ERROR "LibSpiel is needed for ENABLE_SPEECH_SYNTHESIS")
+        endif ()
+        SET_AND_EXPOSE_TO_BUILD(USE_SPIEL ON)
```
is this needed? USE_SPIEL is already ON here.

philn replied:
Yes it's needed. Other USE_* cases do the same. This adds the USE_ define in cmakeconfig.h which (afaik) is then used to compute USE(...) macros.

magomez commented on Source/cmake/OptionsGTK.cmake:
```diff
@@ -357,9 +359,20 @@ if (USE_GBM)
+        if (NOT Flite_FOUND)
+            message(FATAL_ERROR "Flite is needed for ENABLE_SPEECH_SYNTHESIS")
+        endif ()
+        SET_AND_EXPOSE_TO_BUILD(USE_FLITE ON)
```
ditto with FLITE

magomez commented on Source/cmake/OptionsWPE.cmake:
```diff
@@ -221,9 +223,20 @@ if (ENABLE_GAMEPAD AND (NOT (WPE_VERSION VERSION_GREATER_EQUAL 1.13.90)))
+        if (NOT LibSpiel_FOUND)
+            message(FATAL_ERROR "LibSpiel is needed for ENABLE_SPEECH_SYNTHESIS")
+        endif ()
+        SET_AND_EXPOSE_TO_BUILD(USE_SPIEL ON)
```
same as the gtk case, is this needed?

magomez commented on Source/cmake/OptionsWPE.cmake:
```diff
@@ -221,9 +223,20 @@ if (ENABLE_GAMEPAD AND (NOT (WPE_VERSION VERSION_GREATER_EQUAL 1.13.90)))
+        if (NOT Flite_FOUND)
+            message(FATAL_ERROR "Flite is needed for ENABLE_SPEECH_SYNTHESIS")
+        endif ()
+        SET_AND_EXPOSE_TO_BUILD(USE_FLITE ON)
```
and the same again

magomez commented on Source/WebCore/PlatformGTK.cmake:
```diff
@@ -121,12 +121,24 @@ if (ENABLE_BUBBLEWRAP_SANDBOX)
+        )
+        list(APPEND WebCore_LIBRARIES
+            ${Flite_LIBRARIES}
+        )
```
PlatformSpeechSynthesizerGStreamer.cpp and WebKitFliteSourceGStreamer.cpp should be added to the build here, only when flite is enabled, and removed from SourcesGStreamer.txt (not sure there are other files that also need to be moved here)

philn replied:
The previous version of this patch guarded that file with USE(FLITE) which is better i think. I'll restore this part.

philn replied:
It's already the case actually :)

magomez replied:
But the behavior between spiel and flite is not coherent. In the spiel case, PlatformSpeechSynthesizerSpiel.cpp is added to the build here only when USE_SPIEL is enabled, even when you have a check inside the file that ensures that spiel is enabled.
On the flite case, the specific files are added unconditionally to the build by having them on SourcesGStreamer.txt. I's true that they have a check for USE(FLITE) inside and they will be empty, but this is not the same thing that we're doing with spiel.

philn replied:
And what is the issue then?

philn replied:
I mean, I can create a SourcesSpiel.txt file, but that would contain only one unit, which seems a bit overkill, TBH.

magomez replied:
you just need to remove PlatformSpeechSynthesizerGStreamer.cpp and WebKitFliteSourceGStreamer.cpp from SourcesGStreamer and add for flite
 list(APPEND WebCore_SOURCES
            platform/gstreamer/PlatformSpeechSynthesizerGStreamer.cpp
            platform/gstreamer/WebKitFliteSourceGStreamer.cpp
)

magomez commented on Source/cmake/GStreamerDependencies.cmake:
```diff
@@ -1,3 +1,4 @@
 WEBKIT_OPTION_DEPEND(USE_GSTREAMER_MPEGTS ENABLE_VIDEO)
 WEBKIT_OPTION_DEPEND(USE_GSTREAMER_TRANSCODER ENABLE_MEDIA_RECORDER)
 WEBKIT_OPTION_DEPEND(USE_GSTREAMER_WEBRTC ENABLE_WEB_RTC)
+WEBKIT_OPTION_DEPEND(USE_FLITE ENABLE_SPEECH_SYNTHESIS)
```
We're missing the
WEBKIT_OPTION_DEPEND(USE_SPIEL ENABLE_SPEECH_SYNTHESIS)
to ensure that USE_SPIEL is disabled if speech synthesis is disabled.

magomez commented on Source/WebCore/PlatformWPE.cmake:
```diff
@@ -142,12 +142,24 @@ if (ENABLE_GAMEPAD)
+        )
+        list(APPEND WebCore_LIBRARIES
+            ${Flite_LIBRARIES}
+        )
```
ditto

calvaris commented on Source/WebCore/platform/spiel/PlatformSpeechSynthesizerSpiel.cpp:
```diff
@@ -0,0 +1,308 @@
+
+    // TODO: Plumb support for boundaryEventOccurred? Using range-started signal?
+
+    g_signal_connect_swapped(m_speaker.get(), "utterance-started", G_CALLBACK(+[](SpielSpeechWrapper* self, SpielUtterance*) {
+        self->m_platformSynthesizer.client().didStartSpeaking(*self->m_utterance);
+    }), this);
+
+    g_signal_connect_swapped(m_speaker.get(), "utterance-finished", G_CALLBACK(+[](SpielSpeechWrapper* self, SpielUtterance*) {
+        self->m_platformSynthesizer.client().didFinishSpeaking(*self->m_utterance);
+        self->clearUtterance();
+    }), this);
+
+    g_signal_connect_swapped(m_speaker.get(), "utterance-canceled", G_CALLBACK(+[](SpielSpeechWrapper* self, SpielUtterance*) {
+        self->m_platformSynthesizer.client().didFinishSpeaking(*self->m_utterance);
+        self->clearUtterance();
+    }), this);
+
+    g_signal_connect_swapped(m_speaker.get(), "utterance-error", G_CALLBACK(+[](SpielSpeechWrapper* self, SpielUtterance*) {
+        self->m_platformSynthesizer.client().speakingErrorOccurred(*self->m_utterance);
+        self->clearUtterance();
+    }), this);
```
Shouldn't we at least ASSERT on the uterance passed to the callback the same that we have kept in `m_utterance`? I don't know if this is even possible, but it would make sense.

philn replied:
We could compare the utterance text with the one in m_utterrance (which type's different from `SpielUtterance`) ... not sure how useful that is though.

calvaris replied:
Leave it then.

calvaris commented on Source/cmake/OptionsWPE.cmake:
```diff
@@ -96,10 +96,12 @@ WEBKIT_OPTION_DEFINE(ENABLE_WPE_PLATFORM_WAYLAND "Whether to enable support for
 WEBKIT_OPTION_DEFINE(USE_LIBBACKTRACE "Whether to enable usage of libbacktrace." PUBLIC ON)
 WEBKIT_OPTION_DEFINE(USE_LIBDRM "Whether to enable usage of libdrm." PUBLIC ON)
 WEBKIT_OPTION_DEFINE(USE_QT6 "Whether to use Qt6 instead of Qt5" PUBLIC OFF)
+WEBKIT_OPTION_DEFINE(USE_SPIEL "Whether to enable usage of LibSpiel for speech synthesis." PUBLIC OFF)
```
First, I think I would make libspiel private for now as we are still wondering if it's viable because of lack of packetization, both for WPE and GTK.

Second, isn't there a GStreamer files we can use for this and avoid the duplication of options, I remember there was something

philn committed:
[WPE][GTK] Speech synthesis using libspiel
https://bugs.webkit.org/show_bug.cgi?id=270742

Reviewed by Xabier Rodriguez-Calvar and Miguel Gomez.

Spiel's API is much nicer than Flite's, being directly inspired from WebSpeech. This new backend can
be enabled by setting USE_SPIEL=ON and USE_FLITE=OFF with cmake. The flite backend remains enabled
by default until Spiel becomes widly available in the various distributions.

* Source/WebCore/PlatformGTK.cmake:
* Source/WebCore/PlatformWPE.cmake:
* Source/WebCore/platform/PlatformSpeechSynthesizer.h:
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::registerWebKitGStreamerElements):
* Source/WebCore/platform/gstreamer/GUniquePtrFlite.h:
* Source/WebCore/platform/gstreamer/PlatformSpeechSynthesizerGStreamer.cpp:
* Source/WebCore/platform/gstreamer/WebKitFliteSourceGStreamer.cpp:
* Source/WebCore/platform/gstreamer/WebKitFliteSourceGStreamer.h:
* Source/WebCore/platform/spiel/PlatformSpeechSynthesizerSpiel.cpp: Added.
(WebCore::SpielSpeechWrapper::SpielSpeechWrapper):
(WebCore::SpielSpeechWrapper::finishSpeakerInitialization):
(WebCore::SpielSpeechWrapper::~SpielSpeechWrapper):
(WebCore::SpielSpeechWrapper::generateVoiceURI):
(WebCore::SpielSpeechWrapper::initializeVoiceList):
(WebCore::SpielSpeechWrapper::pause):
(WebCore::SpielSpeechWrapper::resume):
(WebCore::SpielSpeechWrapper::speakUtterance):
(WebCore::SpielSpeechWrapper::cancel):
(WebCore::SpielSpeechWrapper::clearUtterance):
(WebCore::PlatformSpeechSynthesizer::create):
(WebCore::PlatformSpeechSynthesizer::PlatformSpeechSynthesizer):
(WebCore::PlatformSpeechSynthesizer::initializeVoiceList):
(WebCore::PlatformSpeechSynthesizer::pause):
(WebCore::PlatformSpeechSynthesizer::resume):
(WebCore::PlatformSpeechSynthesizer::speak):
(WebCore::PlatformSpeechSynthesizer::cancel):
(WebCore::PlatformSpeechSynthesizer::resetState):
* Source/cmake/FindLibSpiel.cmake: Added.
* Source/cmake/GStreamerDependencies.cmake:
* Source/cmake/OptionsGTK.cmake:
* Source/cmake/OptionsWPE.cmake:

Canonical link: https://commits.webkit.org/285891@main
